torch/distributed/_shard/replicated_tensor.py

```diff
@@ -25,66 +26,85 @@ class ReplicatedTensor(torch.Tensor):
-        has_tensor = False
-        for arg in args:
-            if isinstance(arg, torch.Tensor) and not isinstance(arg, ReplicatedTensor):
-                has_tensor = True
-
-        if kwargs is not None:
-            for k, v in kwargs.items():
-                if isinstance(v, torch.Tensor) and not isinstance(v, ReplicatedTensor):
-                    has_tensor = True
-
-        # do dispatch base on the inter-op rules we defined
+        if kwargs is None:
+            kwargs = {}
+        # We will re-dispatch the execution to ShardedTensor __torch_function__
+        # if we find there're ShardedTensor operands. We will also check if args/kwargs
+        # are all replicated tensor operands, we have to do this to ensure we do not
+        # converting results back to ReplicatedTensor if not all operands are replicated.
+        all_replicated = True
+        replicated_pg = None
         for arg in args:
             if isinstance(arg, ShardedTensor):
                 # redispatch to ShardedTensor
-                # TODO: handle ShardedTensor inter-op with ReplicatedTensor
+                # TODO: handle ShardedTensor/PartialTensor inter-op with ReplicatedTensor
                 return arg.__torch_function__(func, types, args, kwargs)
+            if isinstance(arg, ReplicatedTensor):
+                if replicated_pg is None:
+                    replicated_pg = arg.process_group
+                elif replicated_pg != arg.process_group:
+                    raise RuntimeError(
+                        f"ReplicatedTensor operands must be in the same process group "
+                        f"in torch function '{func.__name__}', but found at least two "
+                        f"ReplicatedTensor operands in different process groups! ")
             else:
-                new_args.append(arg)
-
+                all_replicated = False
 
         if kwargs is not None:
             for k, v in kwargs.items():
                 if isinstance(v, ShardedTensor):
                     # redispatch to ShardedTensor
-                    # TODO: handle ShardedTensor inter-op with ReplicatedTensor
+                    # TODO: handle ShardedTensor/PartialTensor inter-op with ReplicatedTensor
                     return v.__torch_function__(func, types, args, kwargs)
+                if isinstance(v, ReplicatedTensor):
+                    if replicated_pg is None:
+                        replicated_pg = v.process_group
+                    elif replicated_pg != v.process_group:
+                        raise RuntimeError(
+                            f"ReplicatedTensor operands must be in the same process group "
+                            f"in torch function '{func.__name__}', but found at least two "
+                            f"ReplicatedTensor operands in different process groups! ")
                 else:
-                    new_kwargs[k] = v
+                    all_replicated = False
```

Comment:
nit: repeated code between args and kwargs, maybe create a simple inline helper function and dedup this.